### PR TITLE
Sprint 4

### PR DESCRIPTION
**Pull request by BIGduzy**
This sprint contains:  [VS2](https://trello.com/c/SSCXmvDL/18-vs2-vertical-slice)

### CODE REVIEW

File: src/controller.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+MotorController::Controller::Controller() : dcInterface(), stepperInterface() {
+}
+
+MotorController::Controller::INTERFACE MotorController::Controller::getSelectedInterface() const {
+    // Getting while interface is nullptr
+    if (selectedInterface == nullptr) {
+        hwlib::cout << "Error:: Trying to get an nullptr interface (controller.cpp, line: " << __LINE__ << ")" << hwlib::endl;
+        // Better not continue
+        while (1) {
```
Comment (JulianvDoorn) [inline src/controller.cpp]: This loop would prevent error handling, (e.g. check if an interface has been selected, otherwise do {some action}), its advised to add an enum item named "None" and return that instead.

File: src/controller.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    return MotorController::Controller::INTERFACE::STEPPER;
+}
+
+void MotorController::Controller::setSelectedInterface(const MotorController::Controller::INTERFACE newInterface) {
+    // Prevent setting the interface a second time
+    if (selectedInterface != nullptr) {
+        return;
+    }
+
+    switch (newInterface) {
```
Comment (JulianvDoorn) [inline src/controller.cpp]: In my opinion, a switch statement is dangerous (forgetting breaks, which results in unpredictable behavior and may cost a lot of time to debug)
Use ifs and elseifs otherwise.

File: src/controller.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        selectedInterface = &dcInterface;
+        break;
+    case MotorController::Controller::INTERFACE::STEPPER:
+        selectedInterface = &stepperInterface;
+        break;
+    default:
+        // Interface not supported
+        hwlib::cout << "Error: Trying to set and not supported interface (controller.cpp, line: " << __LINE__ << ")" << hwlib::endl;
+        // Better not continue
+        while (1) {
```
Comment (JulianvDoorn) [inline src/controller.cpp]: Instead of blocking further execution, `return false;` may be preferred.

File: src/controller.hpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+ * @brief Controller for motor logic
+ *
+ * This class is used to control a motor using the motor_interface superclass.
+ * The interface can be set to DC or STEPPER and uses the selected interface to set speed, direction or angle of a motor
+ *
+ */
+class Controller {
+  public:
+    ///< Enum class with all available motor interfaces
+    enum class INTERFACE { DC, STEPPER };
```
Comment (JulianvDoorn) [inline src/controller.hpp]: Personally, an enum class should be named PascalCases, just like classnames. Fully caps names for macros are introduced due to macros being dangerous in C and required to be used with care. We are programming in C++ where the usage of macros almost never is required, so there is no particular reason for them to be fully upper case in R2D2.

File: src/dc_interface.cpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    }
+}
+
+uint16_t MotorController::DcInterface::getAngle() const {
+    return angle;
+}
+
+void MotorController::DcInterface::setAngle(const uint16_t newAngle) {
+    // Speed can only be 0 - 360
+    if (newAngle <= 360) {
```
Comment (JulianvDoorn) [inline src/dc_interface.cpp]: 360 equals 0 degrees. So it could be better that less-than is used over less-than-or-equal. Perhaps a modulo could be ran over newAngle, but it can be argued that the module using the motor controller should be responsible for this.

**Review by JulianvDoorn — CHANGES_REQUESTED** [review]
Everything is good, except for:
- Some personal flavor differences
- `while (1) {}` is used for error resolution (which would cause rtos to crash, if it was implemented) when false or null can be returned for easier identification for errors. If `while (1) {}` was decided to be used for safety reasons (e.g. stop the motor at all times), a global 'safety' flag could be set instead, which would be required to be unset for motors to function.
Tl;dr running `while (1) {}` in erroneous state just makes matters even worse

**Comment by BendeguzToth** [comment]
@JulianvDoorn Implemented proposed changes.
